### PR TITLE
Remove trailing zeros from Decimal serialization #15794

### LINKED ISSUE #15794: Do not write trailing zeros for decimal values in clickhouse-client or TSV/CSV format
**Use case**
Pretty print and export decimal values which have high accuracy.

**Describe the solution you'd like**
Have a setting which would remove trailing zeros from clickhouse-client and TSV/CSV format output of decimal values.

**Describe alternatives you've considered**
```
taxi9 :) SELECT trim(TRAILING '0' FROM toString(toDecimal128('1312312231.1',20))) as s;

SELECT replaceRegexpOne(toString(toDecimal128('1312312231.1', 20)), concat('[', regexpQuoteMeta('0'), ']*$'), '') AS s

┌─s────────────┐
│ 1312312231.1 │
└──────────────┘

1 rows in set. Elapsed: 0.002 sec.

```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Backward Incompatible Change

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not output trailing zeros in text representation of `Decimal` types. Example: `1.23` will be printed instead of `1.230000` for decimal with scale 6. This closes #15794. It may introduce slight incompatibility if your applications somehow relied on the trailing zeros. Serialization in output formats can be controlled with the setting `output_format_decimal_trailing_zeros`. Implementation of `toString` and casting to String is changed unconditionally.

### CODE REVIEW

**Comment by filimonov** [comment]
> It may introduce slight incompatibility if your applications somehow relied on the trailing zeros.

Having an option for that would be safer. I know people who cast to decimal just to add trailing zeros (BTW there is no nice other  way to do that)

**Comment by alexey-milovidov** [comment]
Something like custom number formatting function...

**Comment by alexey-milovidov** [comment]
Maybe there is a chance to make it under setting.

**Comment by alexey-milovidov** [comment]
Ok, I will make a setting for it.

**Comment by akuzm** [comment]
We use decimals to get fixed-point, fixed-width formatting in performance test reports. AFAIK there is no other way to do this in ClickHouse. I added the setting to the test script.

**Comment by alexey-milovidov** [comment]
Ok, the tests have passed.

I will also extend the test to cover the new setting and additionally validate performance.

**Comment by alexey-milovidov** [comment]
There is some slight performance degradation (or I comparing the build without LTO on my machine with the official build with LTO):

```
SELECT count() FROM system.numbers WHERE NOT ignore(toString((rand() / 1000000)::Decimal64(6)))

Before:

36.47 million rows/s.

  30,56%  clickhouse                         [.] DB::writeText<long>
   4,37%  clickhouse                         [.] DB::WriteBuffer::write
   4,26%  clickhouse                         [.] impl::convert::uitoa<unsigned long, 8ul>
   3,67%  clickhouse                         [.] DB::convertToDecimalImpl<DB::DataTypeNumber<double>, DB::DataTypeDecimal<DB::Decimal<long> >, void>
   3,60%  clickhouse                         [.] DB::ConvertImpl<DB::DataTypeDecimal<DB::Decimal<long> >, DB::DataTypeString, DB::NameToString, DB::ConvertDefaultBehaviorTag>::execute
   2,50%  clickhouse                         [.] DB::impl_::BinaryOperation<unsigned int, unsigned int, DB::DivideFloatingImpl<unsigned int, unsigned int>, double>::process<(DB::impl_::OpCase)2>
   1,72%  clickhouse                         [.] memcpy

After:
   
34.65 million rows/s.
   
  32,50%  clickhouse                         [.] DB::writeText<long>
   4,60%  clickhouse                         [.] DB::WriteBuffer::write
   4,06%  clickhouse                         [.] impl::convert::uitoa<unsigned long, 8ul>
   3,49%  clickhouse                         [.] DB::convertToDecimalImpl<DB::DataTypeNumber<double>, DB::DataTypeDecimal<DB::Decimal<long> >, void>
   2,98%  clickhouse                         [.] DB::ConvertImpl<DB::DataTypeDecimal<DB::Decimal<long> >, DB::DataTypeString, DB::NameToString, DB::ConvertDefaultBehaviorTag>::execute
   2,52%  clickhouse                         [.] DB::impl_::BinaryOperation<unsigned int, unsigned int, DB::DivideFloatingImpl<unsigned int, unsigned int>, double>::process<(DB::impl_::OpCase)2>
   1,71%  clickhouse                         [.] memcpy
```

**Comment by alexey-milovidov** [comment]
I will add performance test.

File: src/IO/WriteHelpers.h [hunk 1]
```diff
@@ -901,34 +901,67 @@ inline void writeText(const LocalDateTime & x, WriteBuffer & buf) { writeDateTim
 inline void writeText(const UUID & x, WriteBuffer & buf) { writeUUIDText(x, buf); }
 
 template <typename T>
-String decimalFractional(const T & x, UInt32 scale)
+void writeDecimalFractional(const T & x, UInt32 scale, WriteBuffer & ostr, bool trailing_zeros)
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:904]: This is the only part of code worth reviewing.

**Comment by filimonov** [comment]
> We use decimals to get fixed-point, fixed-width formatting in performance test reports. AFAIK there is no other way to do this in ClickHouse. I added the setting to the test script.

Maybe it should be added?  In mysql you can do it with rpad/lpad, or with some formatting functions.

That would make life of people who currently may rely on decimal formatting rules (also via toString) easier.

**Comment by alexey-milovidov** [comment]
Yes, I'm already thinking how better to add it.

We already have:
- `format` function that is using Python style substitutions in form of `{}` or `{N}` to paste substrings into a string;
- `formatDateTime` function that is using %-style substitutions to paste datetime components into a string;

We can introduce `formatNumber` function, but I cannot decide between use of fmt-style substitutions and printf-style and between one argument and multiple arguments and possible unification with `format` function that will make it inconsistent with `formatDateTime` and difficult to implement (`fmt` library can be partially reused, but not quite).

So, the better alternative is to implement something like `toStringFixed` function to format only one float or decimal number to string with fixed number of digits. But it is to be confused with FixedString data type. So, we cannot use this name. Maybe name it `toStringWithFixedDigitsAfterDecimal` to be as clear as possible.

**Comment by alexey-milovidov** [comment]
Let it be named `formatDecimal` and take one parameter - the Decimal number.

**Comment by alexey-milovidov** [comment]
Performance test shows speedup for 1.2%:

```
decimal_format | -1.012x |  
```

**Pull request merged**